In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

file_url = 'https://media.githubusercontent.com/media/musthave-ML10/data_source/main/fraud.csv'
data = pd.read_csv(file_url)

data

trans_date_trans_time            cc_num  \
0         2019-01-01 00:00:18  2703186189652095   
1         2019-01-01 00:00:44      630423337322   
2         2019-01-01 00:00:51    38859492057661   
3         2019-01-01 00:01:16  3534093764340240   
4         2019-01-01 00:03:06   375534208663984   
...                       ...               ...   
1852389   2020-12-31 23:59:07    30560609640617   
1852390   2020-12-31 23:59:09  3556613125071656   
1852391   2020-12-31 23:59:15  6011724471098086   
1852392   2020-12-31 23:59:24     4079773899158   
1852393   2020-12-31 23:59:34  4170689372027579   

                                   merchant        category     amt  \
0                fraud_Rippin, Kub and Mann        misc_net    4.97   
1           fraud_Heller, Gutmann and Zieme     grocery_pos  107.23   
2                      fraud_Lind-Buckridge   entertainment  220.11   
3        fraud_Kutch, Hermiston and Farrell   gas_transport   45.00   
4                       fraud_Keeling-Crist        misc_pos   41.96   
...                                     ...             ...     ...   
1852389               fraud_Reilly and Sons  health_fitness   43.77   
1852390                fraud_Hoppe-Parisian       kids_pets  111.84   
1852391                     fraud_Rau-Robel       kids_pets   86.88   
1852392               fraud_Breitenberg LLC          travel    7.99   
1852393                   fraud_Dare-Marvin   entertainment   38.13   

             first     last gender                        street  \
0         Jennifer    Banks      F                561 Perry Cove   
1        Stephanie     Gill      F  43039 Riley Greens Suite 393   
2           Edward  Sanchez      M      594 White Dale Suite 530   
3           Jeremy    White      M   9443 Cynthia Court Apt. 038   
4            Tyler   Garcia      M              408 Bradley Rest   
...            ...      ...    ...                           ...   
1852389    Michael    Olson      M           558 Michael Estates   
1852390       Jose  Vasquez      M           572 Davis Mountains   
1852391        Ann   Lawson      F    144 Evans Islands Apt. 683   
1852392       Eric  Preston      M    7020 Doyle Stream Apt. 951   
1852393     Samuel     Frey      M      830 Myers Plaza Apt. 384   

                   city  ...      lat      long  city_pop  \
0        Moravian Falls  ...  36.0788  -81.1781      3495   
1                Orient  ...  48.8878 -118.2105       149   
2            Malad City  ...  42.1808 -112.2620      4154   
3               Boulder  ...  46.2306 -112.1138      1939   
4              Doe Hill  ...  38.4207  -79.4629        99   
...                 ...  ...      ...       ...       ...   
1852389           Luray  ...  40.4931  -91.8912       519   
1852390    Lake Jackson  ...  29.0393  -95.4401     28739   
1852391         Burbank  ...  46.1966 -118.9017      3684   
1852392            Mesa  ...  44.6255 -116.4493       129   
1852393          Edmond  ...  35.6665  -97.4798    116001   

                                       job         dob  \
0                Psychologist, counselling  1988-03-09   
1        Special educational needs teacher  1978-06-21   
2              Nature conservation officer  1962-01-19   
3                          Patent attorney  1967-01-12   
4           Dance movement psychotherapist  1986-03-28   
...                                    ...         ...   
1852389                       Town planner  1966-02-13   
1852390                     Futures trader  1999-12-27   
1852391                           Musician  1981-11-29   
1852392                       Cartographer  1965-12-15   
1852393                        Media buyer  1993-05-10   

                                trans_num   unix_time  merch_lat  merch_long  \
0        0b242abb623afc578575680df30655b9  1325376018  36.011293  -82.048315   
1        1f76529f8574734946361c461b024d99  1325376044  49.159047 -118.186462   
2        a1a22d70485983eac12b5b88dad1cf95  1325376051  

In [2]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1852394 entries, 0 to 1852393
Data columns (total 22 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   trans_date_trans_time  object 
 1   cc_num                 int64  
 2   merchant               object 
 3   category               object 
 4   amt                    float64
 5   first                  object 
 6   last                   object 
 7   gender                 object 
 8   street                 object 
 9   city                   object 
 10  state                  object 
 11  zip                    int64  
 12  lat                    float64
 13  long                   float64
 14  city_pop               int64  
 15  job                    object 
 16  dob                    object 
 17  trans_num              object 
 18  unix_time              int64  
 19  merch_lat              float64
 20  merch_long             float64
 21  is_fraud               int64  
dtypes: float64(5), int

In [3]:
origin_data = data.copy()
origin_data

trans_date_trans_time            cc_num  \
0         2019-01-01 00:00:18  2703186189652095   
1         2019-01-01 00:00:44      630423337322   
2         2019-01-01 00:00:51    38859492057661   
3         2019-01-01 00:01:16  3534093764340240   
4         2019-01-01 00:03:06   375534208663984   
...                       ...               ...   
1852389   2020-12-31 23:59:07    30560609640617   
1852390   2020-12-31 23:59:09  3556613125071656   
1852391   2020-12-31 23:59:15  6011724471098086   
1852392   2020-12-31 23:59:24     4079773899158   
1852393   2020-12-31 23:59:34  4170689372027579   

                                   merchant        category     amt  \
0                fraud_Rippin, Kub and Mann        misc_net    4.97   
1           fraud_Heller, Gutmann and Zieme     grocery_pos  107.23   
2                      fraud_Lind-Buckridge   entertainment  220.11   
3        fraud_Kutch, Hermiston and Farrell   gas_transport   45.00   
4                       fraud_Keeling-Crist        misc_pos   41.96   
...                                     ...             ...     ...   
1852389               fraud_Reilly and Sons  health_fitness   43.77   
1852390                fraud_Hoppe-Parisian       kids_pets  111.84   
1852391                     fraud_Rau-Robel       kids_pets   86.88   
1852392               fraud_Breitenberg LLC          travel    7.99   
1852393                   fraud_Dare-Marvin   entertainment   38.13   

             first     last gender                        street  \
0         Jennifer    Banks      F                561 Perry Cove   
1        Stephanie     Gill      F  43039 Riley Greens Suite 393   
2           Edward  Sanchez      M      594 White Dale Suite 530   
3           Jeremy    White      M   9443 Cynthia Court Apt. 038   
4            Tyler   Garcia      M              408 Bradley Rest   
...            ...      ...    ...                           ...   
1852389    Michael    Olson      M           558 Michael Estates   
1852390       Jose  Vasquez      M           572 Davis Mountains   
1852391        Ann   Lawson      F    144 Evans Islands Apt. 683   
1852392       Eric  Preston      M    7020 Doyle Stream Apt. 951   
1852393     Samuel     Frey      M      830 Myers Plaza Apt. 384   

                   city  ...      lat      long  city_pop  \
0        Moravian Falls  ...  36.0788  -81.1781      3495   
1                Orient  ...  48.8878 -118.2105       149   
2            Malad City  ...  42.1808 -112.2620      4154   
3               Boulder  ...  46.2306 -112.1138      1939   
4              Doe Hill  ...  38.4207  -79.4629        99   
...                 ...  ...      ...       ...       ...   
1852389           Luray  ...  40.4931  -91.8912       519   
1852390    Lake Jackson  ...  29.0393  -95.4401     28739   
1852391         Burbank  ...  46.1966 -118.9017      3684   
1852392            Mesa  ...  44.6255 -116.4493       129   
1852393          Edmond  ...  35.6665  -97.4798    116001   

                                       job         dob  \
0                Psychologist, counselling  1988-03-09   
1        Special educational needs teacher  1978-06-21   
2              Nature conservation officer  1962-01-19   
3                          Patent attorney  1967-01-12   
4           Dance movement psychotherapist  1986-03-28   
...                                    ...         ...   
1852389                       Town planner  1966-02-13   
1852390                     Futures trader  1999-12-27   
1852391                           Musician  1981-11-29   
1852392                       Cartographer  1965-12-15   
1852393                        Media buyer  1993-05-10   

                                trans_num   unix_time  merch_lat  merch_long  \
0        0b242abb623afc578575680df30655b9  1325376018  36.011293  -82.048315   
1        1f76529f8574734946361c461b024d99  1325376044  49.159047 -118.186462   
2        a1a22d70485983eac12b5b88dad1cf95  1325376051  

In [4]:
data.info(show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1852394 entries, 0 to 1852393
Data columns (total 22 columns):
 #   Column                 Non-Null Count    Dtype  
---  ------                 --------------    -----  
 0   trans_date_trans_time  1852394 non-null  object 
 1   cc_num                 1852394 non-null  int64  
 2   merchant               1852394 non-null  object 
 3   category               1852394 non-null  object 
 4   amt                    1852394 non-null  float64
 5   first                  1852394 non-null  object 
 6   last                   1852394 non-null  object 
 7   gender                 1852394 non-null  object 
 8   street                 1852394 non-null  object 
 9   city                   1852394 non-null  object 
 10  state                  1852394 non-null  object 
 11  zip                    1852394 non-null  int64  
 12  lat                    1852394 non-null  float64
 13  long                   1852394 non-null  float64
 14  city_pop          

In [5]:
round(data.describe(), 2)

cc_num         amt         zip         lat        long  \
count  1.852394e+06  1852394.00  1852394.00  1852394.00  1852394.00   
mean   4.173860e+17       70.06    48813.26       38.54      -90.23   
std    1.309115e+18      159.25    26881.85        5.07       13.75   
min    6.041621e+10        1.00     1257.00       20.03     -165.67   
25%    1.800429e+14        9.64    26237.00       34.67      -96.80   
50%    3.521417e+15       47.45    48174.00       39.35      -87.48   
75%    4.642255e+15       83.10    72042.00       41.94      -80.16   
max    4.992346e+18    28948.90    99921.00       66.69      -67.95   

         city_pop     unix_time   merch_lat  merch_long    is_fraud  
count  1852394.00  1.852394e+06  1852394.00  1852394.00  1852394.00  
mean     88643.67  1.358674e+09       38.54      -90.23        0.01  
std     301487.62  1.819508e+07        5.11       13.76        0.07  
min         23.00  1.325376e+09       19.03     -166.67        0.00  
25%        741.00  1.343017e+09       34.74      -96.90        0.00  
50%       2443.00  1.357089e+09       39.37      -87.44        0.00  
75%      20328.00  1.374581e+09       41.96      -80.25        0.00  
max    2906700.00  1.388534e+09       67.51      -66.95        1.00

In [6]:
data.drop(['first','last','street','city','state','zip','trans_num','unix_time','job','merchant'], axis=1, inplace= True)

In [7]:
data['trans_date_trans_time'] = pd.to_datetime(data['trans_date_trans_time']) # 날짜 형식으로 변환

In [8]:
data.info()

round(data.describe(), 2)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1852394 entries, 0 to 1852393
Data columns (total 12 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   trans_date_trans_time  datetime64[ns]
 1   cc_num                 int64         
 2   category               object        
 3   amt                    float64       
 4   gender                 object        
 5   lat                    float64       
 6   long                   float64       
 7   city_pop               int64         
 8   dob                    object        
 9   merch_lat              float64       
 10  merch_long             float64       
 11  is_fraud               int64         
dtypes: datetime64[ns](1), float64(5), int64(3), object(3)
memory usage: 169.6+ MB


trans_date_trans_time        cc_num         amt         lat  \
count                        1852394  1.852394e+06  1852394.00  1852394.00   
mean   2020-01-20 21:31:46.801827328  4.173860e+17       70.06       38.54   
min              2019-01-01 00:00:18  6.041621e+10        1.00       20.03   
25%    2019-07-23 04:13:43.750000128  1.800429e+14        9.64       34.67   
50%              2020-01-02 01:15:31  3.521417e+15       47.45       39.35   
75%    2020-07-23 12:11:25.249999872  4.642255e+15       83.10       41.94   
max              2020-12-31 23:59:34  4.992346e+18    28948.90       66.69   
std                              NaN  1.309115e+18      159.25        5.07   

             long    city_pop   merch_lat  merch_long    is_fraud  
count  1852394.00  1852394.00  1852394.00  1852394.00  1852394.00  
mean       -90.23    88643.67       38.54      -90.23        0.01  
min       -165.67       23.00       19.03     -166.67        0.00  
25%        -96.80      741.00       34.74      -96.90        0.00  
50%        -87.48     2443.00       39.37      -87.44        0.00  
75%        -80.16    20328.00       41.96      -80.25        0.00  
max        -67.95  2906700.00       67.51      -66.95        1.00  
std         13.75   301487.62        5.11       13.76        0.07

trans_date_trans_time 이 object가 아닌 datatime64[ns]로 변경된 것을 확인할 수 있다.

In [9]:
amt_info = data.groupby('cc_num').agg(['mean', 'std'])['amt'].reset_index()

TypeError: agg function failed [how->mean,dtype->object]

In [ ]:
amt_info = data.groupby('cc_num').mean(numeric_only=True)
amt_info = amt_info.drop(['lat','long', 'city_pop', 'merch_lat', 'merch_long', 'is_fraud'], axis=1)
amt_info_std = data.groupby('cc_num').std(numeric_only=True).drop(['lat','long', 'city_pop', 'merch_lat', 'merch_long', 'is_fraud'], axis=1)

amt_info['std'] = amt_info_std['amt']
amt_info['mean'] = amt_info['amt']
amt_info = amt_info.drop('amt', axis=1)

amt_info.head()

In [ ]:
data = data.merge(amt_info, on='cc_num', how='left')

data

In [ ]:
# data = data.drop(['std_x', 'std_y', 'mean_x', 'mean_y'], axis=1)

# data

이제 기존 데이터에 특정값(amt), 평균(mean), 표준편차(std)가 모두 모였으니 Z 점수를 계산.

z_score = 특정값 - 평균 / 표준편차<br/>

특정값이 정규분포 범위에서 어느 수준에 위치하는지를 나타낸다. (z-점수는 자료가 평균으로부터 표준편차의 몇 배만큼 떨어져 있는지를 보여줍니다.)

In [ ]:
data['amt_z_score'] = (data['amt'] - data['mean']) / data['std'] # z-score 계산

In [ ]:
data[['amt', 'mean', 'std', 'amt_z_score']].head()

index 2행을 확인하면 평균 거래 금액이 약 \$69.9 수준인데 해당 지출금액은 \$220로 상당한 차이를 보인다.

In [ ]:
data.drop(['mean', 'std'], axis=1, inplace=True) # 변수 제거

In [ ]:
data

### 11.4.2 범주

In [ ]:
category_info = data.groupby(['cc_num', 'category']).mean(numeric_only=True)
category_info = category_info.drop(['lat', 'long', 'city_pop', 'merch_lat', 'merch_long', 'is_fraud', 'amt_z_score'], axis=1)
category_info_std = data.groupby(['cc_num', 'category']).std(numeric_only=True).drop(['lat','long', 'city_pop', 'merch_lat', 'merch_long', 'is_fraud', 'amt_z_score'], axis=1)

category_info['std'] = category_info_std['amt']
category_info['mean'] = category_info['amt']
category_info = category_info.drop('amt', axis=1)

category_info.head()

In [ ]:
data = data.merge(category_info, on=['cc_num', 'category'], how='left')

In [ ]:
data

In [ ]:
data['cat_z_score'] = (data['amt'] - data['mean']) / data['std']
data.drop(['mean', 'std'], axis=1, inplace=True)

### 11.4.3 거리

두 지점의 거리 계산에 geopy 라이브러리를 사용.

In [ ]:
!pip install geopy

In [ ]:
import geopy.distance

In [ ]:
data['merch_coord'] = pd.Series(zip(data['merch_lat'], data['merch_long']))

data['cust_coord'] = pd.Series(zip(data['lat'], data['long']))

data

In [ ]:
data['distance'] = data.apply(lambda x : geopy.distance.distance(x['merch_coord'], x['cust_coord']).km, axis=1)

In [ ]:
data

In [ ]:
data

In [ ]:
distance_info = data.groupby(['cc_num']).mean(numeric_only=True)
distance_info = distance_info.drop(['lat', 'long', 'city_pop', 'merch_lat', 'merch_long', 'is_fraud', 'amt_z_score', 'cat_z_score', 'distance'], axis=1)
distance_info_std = data.groupby(['cc_num']).std(numeric_only=True).drop(['lat','long', 'city_pop', 'merch_lat', 'merch_long', 'is_fraud', 'amt_z_score', 'cat_z_score', 'distance'], axis=1)

distance_info['std'] = distance_info_std['amt']
distance_info['mean'] = distance_info['amt']
distance_info = distance_info.drop('amt', axis=1)

distance_info.head()

In [ ]:
data = data.merge(distance_info, on='cc_num', how='left')
data['distance_z_score'] = (data['distance']-data['mean'])/data['std']
data.drop(['mean', 'std'], axis=1, inplace=True)
data.head()

In [ ]:
data['age'] = 2021 - pd.to_datetime(data['dob']).dt.year # 나이 계산
data

In [ ]:
data.drop(['cc_num', 'lat', 'long', 'merch_lat', 'merch_long', 'dob', 'merch_coord', 'cust_coord'], axis=1, inplace=True)

data.head()

In [ ]:
data = pd.get_dummies(data, columns=['category', 'gender'], drop_first=True)

In [ ]:
data.set_index('trans_date_trans_time', inplace=True)
data.head()

In [ ]:
train = data[data.index < '2020-07-01']
test = data[data.index >= '2020-07-01']

In [ ]:
len(test) / len(data)

In [ ]:
x_train = train.drop('is_fraud', axis=1)
x_test = test.drop('is_fraud', axis=1)
y_train = train['is_fraud']

In [ ]:
y_test = test['is_fraud']

In [ ]:
import lightgbm as lgb

In [ ]:
model_1 = lgb.LGBMClassifier(random_state=100)
model_1.fit(x_train, y_train)
pred_1 = model_1.predict(x_test)

In [ ]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report, roc_auc_score

In [ ]:
accuracy_score(y_test, pred_1)

In [ ]:
print(confusion_matrix(y_test, pred_1))

In [ ]:

print(classification_report(y_test, pred_1))

In [ ]:
proba_1 = model_1.predict_proba(x_test)

In [ ]:
proba_1

In [ ]:
proba_1 = proba_1[:, 1]

In [ ]:
proba_int1 = (proba_1 > 0.2).astype('int')
proba_int2 = (proba_1 > 0.8).astype('int')

In [ ]:
print(confusion_matrix(y_test, proba_int1))

In [ ]:
print(classification_report(y_test, proba_int1))

In [ ]:
print(confusion_matrix(y_test, proba_int2))
print(classification_report(y_test, proba_int2))

In [ ]:
roc_auc_score(y_test, proba_1)

In [ ]:
from sklearn.model_selection import RandomizedSearchCV

params = {
    'n_estimators' : [100, 500, 1000],
    'learning_rate' : [0.01, 0.05, 0.1, 0.3],
    'lambda_l1': [0, 10, 20, 30, 50],
    'lambda_l2': [0, 10, 20, 30, 50],
    'max_depth': [5, 10, 15, 20],
    'subsample': [0.6, 0.8, 1]
}

In [ ]:
model_2 = lgb.LGBMClassifier(random_state=100)
rs = RandomizedSearchCV(model_2, param_distributions=params, n_iter=30, scoring='roc_auc', random_state=100, n_jobs=-1)

In [ ]:
import time
start = time.time()
rs.fit(x_train, y_train)
print(time.time() - start)

In [ ]:
rs.best_params_

In [ ]:
rs_proba = rs.predict_proba(x_test)
roc_auc_score(y_test, rs_proba[:, 1])

In [ ]:
rs_proba_int = (rs_proba[:, 1] > 0.2).astype('int')
print(confusion_matrix(y_test, rs_proba_int))

In [ ]:
print(classification_report(y_test, rs_proba_int))

In [ ]:

train = data[data.index < '2020-01-01']
val = data[(data.index >= '2020-01-01') & (data.index < '2020-07-01')]
test = data[data.index >= '2020-07-01']

In [ ]:
x_train = train.drop('is_fraud', axis=1)
x_val = val.drop('is_fraud', axis=1)
x_test = test.drop('is_fraud', axis=1)
y_train = train['is_fraud']
y_val = val['is_fraud']
y_test = test['is_fraud']

In [ ]:
d_train = lgb.Dataset(x_train, label=y_train)
d_val = lgb.Dataset(x_val, label=y_val)

In [ ]:
params_set = rs.best_params_
params_set['metrics'] = 'auc'

In [ ]:
params_set

In [ ]:
model_3 = lgb.train(params_set, d_train, valid_sets=[d_val], early_stopping_rounds=100, verbose_eval=100)

In [ ]:
pred_3 = model_3.predict(x_test)

In [ ]:
roc_auc_score(y_test, pred_3)

In [ ]:
feature_imp = pd.DataFrame({'feature_name' : x_train.columns, 'importance': model_1.feature_importances_}).sort_values('importance', ascending=False)

plt.figure(figsize=(20, 10))
sns.barplot(x="importance", y="feature_name", data=feature_imp.head(10))
plt.show()

In [ ]:
feature_imp3 = pd.DataFrame(sorted(zip(model_3.feature_importance(), x_train.columns)), columns=['Value', 'Feature'])

plt.figure(figsize=(20, 10))
sns.barplot(x="importance", y="feature_name", data=feature_imp.head(10))
plt.show()